compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNotImplementedOverrideChecker.kt

```diff
@@ -0,0 +1,278 @@
+        firClass.declarations.filterIsInstance<FirCallableDeclaration<*>>() +
+                dataClassMembers(firClass, context) +
+                delegatedMembers(firClass, context)
+
+    private val FirSimpleFunction.matchesEqualsSignature: Boolean
+        get() = valueParameters.size == 1 && valueParameters[0].returnTypeRef.coneType.isNullableAny &&
+                returnTypeRef.coneType.isBoolean
+
+    private val FirSimpleFunction.matchesHashCodeSignature: Boolean
+        get() = valueParameters.isEmpty() &&
+                returnTypeRef.coneType.isInt
+
+    private val FirSimpleFunction.matchesToStringSignature: Boolean
+        get() = valueParameters.isEmpty() &&
+                returnTypeRef.coneType.isString
+
+    private val FirSimpleFunction.matchesDataClassSyntheticMemberSignatures: Boolean
+        get() = (this.name == EQUALS_NAME && matchesEqualsSignature) ||
+                (this.name == HASHCODE_NAME && matchesHashCodeSignature) ||
+                (this.name == TOSTRING_NAME && matchesToStringSignature)
+
+    private val SYNTHETIC_NAMES = listOf(EQUALS_NAME, HASHCODE_NAME, TOSTRING_NAME)
+
+    // See [DataClassMembersGenerator#generate]
+    // But, this one doesn't create IR counterparts. We just need what synthetic members would be generated for data/inline classes.
+    private fun dataClassMembers(firClass: FirClass<*>, context: CheckerContext): Collection<FirCallableDeclaration<*>> {
+        if (firClass !is FirRegularClass || (!firClass.isData && !firClass.isInline)) return emptyList()
+
+        val contributedInThisType = firClass.declarations.mapNotNull {
+            if (it is FirSimpleFunction && it.matchesDataClassSyntheticMemberSignatures) {
+                it.name
+            } else
+                null
+        }
+
+        val contributedInSupertypes = mutableMapOf<Name, FirCallableDeclaration<*>>()
+        val classScope = firClass.unsubstitutedScope(context)
+        for (name in SYNTHETIC_NAMES) {
+            classScope.processFunctionsByName(name) {
+                val declaration = it.fir
+                if (declaration.matchesDataClassSyntheticMemberSignatures && declaration.modality != Modality.FINAL) {
+                    contributedInSupertypes.putIfAbsent(declaration.name, declaration)
+                }
+            }
+        }
+
+        val result = mutableListOf<FirCallableDeclaration<*>>()
+        for (name in SYNTHETIC_NAMES) {
+            if (name in contributedInThisType) continue
+            contributedInSupertypes[name]?.let { result.add(it) }
+        }
+        return result
+    }
+
+    // See [DelegatedMemberGenerator#generate]
+    // But, this one doesn't create IR counterparts. We just need what members could be delegated to avoid false fake overrides.
+    private fun delegatedMembers(firClass: FirClass<*>, context: CheckerContext): Collection<FirCallableDeclaration<*>> {
```

Comment:
Should we call this function for all classes?

Reply:
Yeah... but we can't ignore delegated members. Needless to say, without this, we will see many false alarms. It starts with `synthetic` `FirField` check to make sure we have a delegation indeed. Hope that early bail-out is effective enough.

Reply:
I took a look at some relevant diagnostics, and I wonder this is indeed inevitable for the following ones:

* OVERRIDING_FINAL_MEMBER_BY_DELEGATION
* VAR_OVERRIDDEN_BY_VAL_BY_DELEGATION
* RETURN_TYPE_MISMATCH_BY_DELEGATION
* PROPERTY_TYPE_MISMATCH_BY_DELEGATION

Here, I used this util to correctly mimic fake override generation, but eventually, we need to collect what members are delegated, and then compare them against inherited members from super types.